### PR TITLE
feat(server-auth-testing): add JUnit 5 support for opaRule and authRule

### PR DESCRIPTION
**Pull request by ericrath**

### CODE REVIEW

File: sda-commons-server-auth-testing/README.md [hunk 1]
```diff
@@ -11,11 +11,12 @@ testCompile 'org.sdase.commons:sda-commons-server-auth-testing:<current-version>
...
 
 In an integration test, authentication can be configured using the 
-[`AuthRule`](./src/main/java/org/sdase/commons/server/auth/testing/AuthRule.java) and/or the [`OpaRule`](./src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java).
+[`AuthRule`](./src/main/java/org/sdase/commons/server/auth/testing/AuthRule.java) and/or the [`OpaRule`](./src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java) (Junit 4) or
+[`AuthExtension`](./src/main/java/org/sdase/commons/server/auth/testing/AuthExtension.java) and/or the [`OpaExtension`](./src/main/java/org/sdase/commons/server/opa/testing/OpaExtension.java) (Junit 5).
 
-## Auth Rule
-The `AuthRule` uses the `EnvironmentRule` to create the `AuthConfig` in an environment property called `AUTH_RULE`.
-Therefore the configuration in the test needs to use this property and the application is required to use the 
+## Auth Extension
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/README.md:17]: Please don't replace the documentation of JUnit 4 features but add documentation for JUnit 5 extensions.

File: sda-commons-server-auth-testing/build.gradle [hunk 2]
```diff
@@ -5,12 +5,12 @@ dependencies {
   compile 'com.auth0:java-jwt'
   compile project(':sda-commons-client-jersey-wiremock-testing')
 
-  // the following modules ship with different versions of the swagger-hal library.
-  // since swagger-hal is not needed in these tests, they are excluded from both.
-  testCompile project(':sda-commons-server-swagger'), {
+  testCompile project(':sda-commons-server-openapi'), {
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/build.gradle:8]: Why did you remove the comments?

File: sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/auth/testing/AuthExtension.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import io.dropwizard.Configuration;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.sdase.commons.server.auth.config.AuthConfig;
+import org.sdase.commons.server.testing.Environment;
+
+public class AuthExtension extends AbstractAuth implements BeforeAllCallback, AfterAllCallback {
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/auth/testing/AuthExtension.java:17]: Do we want a common naming convention? According to Wiremock we could change the name to "AuthClassExtension".

File: sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/auth/testing/AuthRule.java [hunk 4]
```diff
@@ -32,8 +21,11 @@
  * AuthRule#auth() a builder} for creation of tokens with optional custom claims. The rule may be
  * used with custom issuer, subject, certificate and public key. Issuer and subject may be
  * customized for each token.
+ *
+ * @deprecated Please migrate to Junit 5 using {@link AuthExtension}
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/auth/testing/AuthRule.java:25]: Don't add deprecations.

File: sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java [hunk 5]
```diff
@@ -1,335 +1,42 @@
...
 import org.junit.rules.RuleChain;
+import org.junit.rules.TestRule;
 import org.junit.runner.Description;
 import org.junit.runners.model.Statement;
-import org.sdase.commons.server.opa.filter.model.OpaResponse;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
+/** @deprecated migrate to Junit 5 and use {@link OpaExtension} */
+@Deprecated
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java:13]: Don't add deprecations.

File: sda-commons-server-auth-testing/src/test/java/org/sdase/commons/server/auth/testing/AuthDisabledIT.java [hunk 6]
```diff
@@ -2,29 +2,29 @@
...
-import org.junit.ClassRule;
-import org.junit.Test;
-import org.junit.rules.RuleChain;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
 import org.sdase.commons.server.auth.testing.test.AuthTestApp;
 import org.sdase.commons.server.auth.testing.test.AuthTestConfig;
 
-public class AuthDisabledIT {
+class AuthDisabledIT {
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/src/test/java/org/sdase/commons/server/auth/testing/AuthDisabledIT.java:16]: I'd do it the other way round: Keep the existing test mostly like it is and create a new test for JUnit 5.

File: sda-commons-server-auth-testing/src/test/java/org/sdase/commons/server/auth/testing/AuthRuleCustomizationJunit4IT.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.core.GenericType;
+import javax.ws.rs.core.Response;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.sdase.commons.server.auth.testing.test.AuthTestApp;
+import org.sdase.commons.server.auth.testing.test.AuthTestConfig;
+
+public class AuthRuleCustomizationJunit4IT {
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/src/test/java/org/sdase/commons/server/auth/testing/AuthRuleCustomizationJunit4IT.java:22]: Same here: Create new test(s) for JUnit 5 and keep the JUnit 4 test as it is.

File: sda-commons-server-auth-testing/README.md [hunk 8]
```diff
@@ -83,7 +83,76 @@ Examples can be found in the [test source branch](./src/test) of this module. Th
...
+# The configuration of the test auth bundle is injected here
+auth: ${AUTH_RULE}
+```
+
+To implement the test, the `AuthExtension` has to be initialized before the `DropwizardAppExtension`:
+
+```java
+class AuthExtensionIT {
+
+  @RegisterExtension public static final AuthExtension AUTH = AuthExtension.builder().build();
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/README.md:129]: Not 100% sure if we also need `@Order`. I think we get a reproducible but random order otherwise.

File: sda-commons-server-auth-testing/README.md [hunk 9]
```diff
@@ -143,3 +212,68 @@ Examples can be found in the [test source branch](./src/test) of this module. Th
 Example with activated AUTH and OPA bundle can be found here:
 - [Example app](./src/test/java/org/sdase/commons/server/opa/testing/test/AuthAndOpaBundleTestApp.java)
 - [Test](./src/test/java/org/sdase/commons/server/opa/testing/AuthAndOpaIT.java)
+
+## OPA Extension
+
+The Junit 5 OPA Extension is build around WireMock. The mock can be configured via the extension.
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/README.md:218]: ... is built ...

File: sda-commons-server-auth-testing/README.md [hunk 10]
```diff
@@ -143,3 +212,68 @@ Examples can be found in the [test source branch](./src/test) of this module. Th
...
+
+Use extension `DropwizardAppExtension` from `org.sdase.commons:sda-commons-server-auth-testing` instead of the native Dropwizard equivalent.
+The native extension is currently not working well with `@RegisterExtension`.
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/README.md:221]: Can be removed. We can use the original class from Dropwizard.

File: sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/auth/testing/AuthRule.java [hunk 11]
```diff
@@ -6,34 +6,23 @@
...
-import org.sdase.commons.server.auth.config.KeyUriType;
 import org.sdase.commons.server.testing.EnvironmentRule;
 
 /**
  * This {@link TestRule} configures a Dropwizard application that uses JWT authentication for
  * integration tests. It provides a local certificate for JWT verification and offers {@link
  * AuthRule#auth() a builder} for creation of tokens with optional custom claims. The rule may be
  * used with custom issuer, subject, certificate and public key. Issuer and subject may be
- * customized for each token.
+ * customized for each token. q
```
Comment (christopher-cudennec) [inline sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/auth/testing/AuthRule.java:23]: Some leftover.

File: sda-commons-server-auth-testing/README.md (hunk 8 above)
Comment (ericrath) [reply sda-commons-server-auth-testing/README.md:129]: we clarified, that we need @order

Comment (christopher-cudennec) [reply sda-commons-server-auth-testing/README.md:129]: Could you please add `@Order` to the two extensions of the example?

**Review by christopher-cudennec — APPROVED** [review]
🚟
